File: java/domain/src/main/java/models/Rally.java
package models;

import java.util.Objects;

public class Rally extends Identifiable<Long> {
    private int engineCapacity;

    public Rally(int engineCapacity) {
        this.engineCapacity = engineCapacity;
    }

    public int getEngineCapacity() {
        return engineCapacity;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Rally)) return false;
        if (!super.equals(o)) return false;
        Rally rally = (Rally) o;
        return engineCapacity == rally.engineCapacity;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), engineCapacity);
    }
}

File: csharp/Domain/Models/Rally.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Domain.Models
{
    public class Rally : Identifiable<long>  
    {
        public int EngineCapacity { get; set; }
        public int NumberOfParticipants { get; set; }   

        public Rally(int engineCapacity)
        {
            EngineCapacity = engineCapacity;
            NumberOfParticipants = 0;
        }

        public Rally(int engineCapacity, int numberOfParticipants)
        {
            NumberOfParticipants = numberOfParticipants;
            EngineCapacity = engineCapacity;
        }

        public override bool Equals(object? obj)
        {
            return obj is Rally rally &&
                   base.Equals(obj) &&
                   Id == rally.Id &&
                   EngineCapacity == rally.EngineCapacity &&
                   NumberOfParticipants == rally.NumberOfParticipants;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(base.GetHashCode(), Id, EngineCapacity, NumberOfParticipants);
        }
    }
}
